socket.1
.\" -*- nroff -*-
.ig
$Header: socket.1[1.13] Wed Sep  9 16:46:30 1992 nickel@cs.tu-berlin.de proposed $
This file is part of socket(1).
Copyright (C) 1992 by Juergen Nickelsen <nickel@cs.tu-berlin.de>
Please read the file COPYRIGHT for further details.
..
.TH SOCKET 1 "Aug  6, 1992"
.SH NAME
socket \- create tcp socket and connect to stdin/out
.SH SYNOPSIS
.B socket
[
.B \-bcfqrvw
]
[
.B \-p
.I command
]
.I host port
.br
.B socket
[
.B \-bcfqrvw
]
[
.B \-p
.I command
]
.B \-s 
[
.B \-l
]
.I port
.SH DESCRIPTION
.B Socket
creates an Internet domain TCP socket and connects it to stdin and stdout.
The 
.I host
argument can be an Internet number in dot-notation (like
``130.149.28.10'') or a domain name. In this case it must be possible
to resolve the name to a valid Internet address with
.IR gethostbyname (3).
The 
.I port
argument can be a port number or a service name which can be mapped to
a port number by
.IR getservbyname (3).
.SH OPTIONS
.TP
.BR "\-b " (background)
The program forks itself into the background, detaches from its
controlling tty, closes the file descriptors associated with the tty,
and changes its current directory to the root directory.
.TP
.BR "\-c " (crlf)
Linefeed characters (LF) are converted to a Carriage Return Linefeed
sequence (CRLF) when written to the socket.
CRLF sequences read from the socket are converted to a single LF.
.TP
.BR "\-f " (fork)
When a server connection has been accepted, a separate process is
forked to handle the connection in background.
.TP
.BR "\-l " (loop)
(only valid with
.BR \-s )
After a connection has been closed,
another connection is accepted.
.TP
.BR "\-p " (program)
The specified 
.I command
is executed for each connection. Its standard input, standard output,
and standard error channels are connected to the socket.
.I Command
can be any shell command since it is passed to \fC/bin/sh\fP.
.TP
.BR "\-q " (quit)
The connection is closed when an end-of-file condition occurs on standard
input.
.TP
.BR "\-r " "(read only)"
No data is read from standard input and written to the socket.
.TP
.BR "\-s " (server)
A server socket is created.
A
.I hostname 
argument is not required.
.TP
.BR "\-v " (verbose)
Messages about connections etc. are issued to stderr.
.TP
.BR "\-w " "(write only)"
No data is read from the socket and written to the standard output.
.TP
.B \-version
.B Socket
prints its version ID and terminates. 
This must be the first argument to have an effect.
.SH EXAMPLES
The command
.IP
\fCsocket -v coma.cs.tu-berlin.de nntp\fP
.LP
connects to the nntp port (port 119) of coma.cs.tu-berlin.de
(130.149.28.10).
.br
The command
.IP
\fCsocket \-sl 3425\fP
.LP
creates a server socket on port 3425 on the local host and waits for a
connection.
After a connection has been closed, a new connection is accepted. 
.br
The command
.IP
\fCsocket \-wslqvp "echo Socket! " 1938\fP
.LP
creates a server socket on port 1938 on the local host and waits for a
connection.
When a connection is accepted, the string "Socket!" is written to the
socket.
No data is read from the socket and written to the \fCfinger\fP
program.
The connection is closed when an end-of-file condition at the standard
output of the program  occurs.
Then a new connection is accepted.
.SH DIAGNOSTICS
Lots of diagnostics for failed system calls.
.TP
.BI "unknown host " host
.IR host 's
address could not be resolved.
.TP
.B Signal \fIsignal\fP caught, exiting
.B Socket
exits on any signal other than SIGTSTP, SIGCONT, SIGCLD, SIGQUIT.
.LP
A non-zero exit code is returned if 
.B socket
terminates due to an error condition or a signal.
.SH SEE ALSO
.BR accept (2),
.BR bind (2),
.BR connect (2), 
.BR socket (2),
.BR gethostbyname (3),
.BR getservbyname (3)
.SH BUGS
\fCsocket \-p\fP terminates due to a SIGPIPE signal when there is more
data from the socket available than the executed program wants to
read.
.LP
Please report any other bugs to the author.
.SH VERSION
This manual page describes Socket\-1.1.
.SH AUTHOR
Juergen Nickelsen <jn@berlin.snafu.de>


CHANGES
-*-text-*-

Socket version 1.2, 27-Jan-2018
  - a new release based on FreeBSD ports patches
  - the source code is now available on GitHub: https://github.com/wosch/socket

User-visible changes in Socket version 1.1:

  - Ported to more Platforms (SGI Irix, IBM AIX, CDC EP/IX, ISC 3.0,
    System V R4, Sequent)

  - fixed getopt()-related bug in main() (opt declared as int now).

  - fixed SIGWINCH-related bug (socket no longer exits on SIGWINCH).

  - Background mode (-b option)

  - -f option to fork child 

  - fixed bug in server mode concerning the "Connection from..."
    message

  - slightly more adequate signal handling and messages



socket.c
/*

$Header: socket.c[1.17] Wed Sep  9 16:23:14 1992 nickel@cs.tu-berlin.de proposed $
This file is part of socket(1).
Copyright (C) 1992 by Juergen Nickelsen <nickel@cs.tu-berlin.de>
Please read the file COPYRIGHT for further details.

*/

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <errno.h>
#include <stdio.h>
#ifdef SEQUENT
#include <strings.h>
#else
#include <string.h>
#endif
#include <stdlib.h>
#include <unistd.h>
#include "globals.h"

/* global variables */
int forkflag = 0 ;		/* server forks on connection */
int serverflag = 0 ;		/* create server socket */
int loopflag = 0 ;		/* loop server */
int verboseflag = 0 ;		/* give messages */
int readonlyflag = 0 ;		/* only read from socket */
int writeonlyflag = 0 ;		/* only write to socket */
int quitflag = 0 ;		/* quit connection on EOF */
int crlfflag = 0 ;		/* socket expects and delivers CRLF */
int backgflag = 0 ;		/* put yourself in background */
int active_socket ;		/* socket with connection */
char *progname ;		/* name of the game */
char *pipe_program = NULL ;	/* program to execute in two-way pipe */

void server A((int port, char *service_name)) ;
void handle_server_connection A((void)) ;
void client A((char *host, int port, char *service_name)) ;
extern void init_signals A((void)) ;
extern void do_io A((void)) ;

int main(argc, argv)
int argc ;
char **argv ;
{
    char *cp ;			/* to point to '/' in argv[0] */
    int opt ;			/* option character */
    int error = 0 ;		/* usage error occurred */
    extern int optind ;		/* from getopt() */
    /* char *host ; */		/* name of remote host */
    int port ;			/* port number for socket */
    char *service_name ;	/* name of service for port */

    /* print version ID if requested */
    if (argv[1] && !strcmp(argv[1], "-version")) {
	puts(so_release()) ;
	exit(0) ;
    }

    /* set up progname for later use */
    progname = argv[0] ;
    if ((cp = strrchr(progname, '/'))) progname = cp + 1 ;

    /* parse options */
    while ((opt = getopt(argc, argv, "bcflp:qrsvw?")) != -1) {
	switch (opt) {
	  case 'f':
	    forkflag = 1 ;
	    break ;
	  case 'c':
	    crlfflag = 1 ;
	    break ;
	  case 'w':
	    writeonlyflag = 1 ;
	    break ;
	  case 'p':
	    pipe_program = argv[optind - 1] ;
	    break ;
	  case 'q':
	    quitflag = 1 ;
	    break ;
	  case 'r':
	    readonlyflag = 1 ;
	    break ;
	  case 's':
	    serverflag = 1 ;
	    break ;
	  case 'v':
	    verboseflag = 1 ;
	    break ;
	  case 'l':
	    loopflag = 1 ;
	    break ;
	  case 'b':
	    backgflag = 1 ;
	    break ;
	  default:
	    error++ ;
	}
    }
    if (error ||		/* usage error? */
	argc - optind + serverflag != 2) { /* number of args ok? */
	usage() ;
	exit(15) ;
    }

    /* check some option combinations */
#define senseless(s1, s2) \
    fprintf(stderr, "It does not make sense to set %s and %s.\n", (s1), (s2))

    if (writeonlyflag && readonlyflag) {
	senseless("-r", "-w") ;
	exit(15) ;
    }
    if (loopflag && !serverflag) {
	senseless("-l", "not -s") ;
	exit(15) ;
    }
    if (backgflag && !serverflag) {
 	senseless("-b", "not -s") ;
 	exit(15) ;
    }
    if (forkflag && !serverflag) {
	senseless("-f", "not -s") ;
    }
 
    /* set up signal handling */
    init_signals() ;

    /* get port number */
    port = resolve_service(argv[optind + 1 - serverflag],
			   "tcp", &service_name) ;
    if (port < 0) {
	fprintf(stderr, "%s: unknown service\n", progname) ;
	exit(5) ;
    }

    /* and go */
    if (serverflag) {
	if (backgflag) {
	    background() ;
	}
	server(port, service_name) ;
    } else {
	client(argv[optind], port, service_name) ;
    }	       
    exit(0) ;
}


void server(port, service_name)
int port ;
char *service_name ;
{
    int socket_handle, alen ;

    /* allocate server socket */
    socket_handle = create_server_socket(port, 1) ;
    if (socket_handle < 0) {
	perror2("server socket") ;
	exit(1) ;
    }
    if (verboseflag) {
	fprintf(stderr, "listening on port %d", port) ;
	if (service_name) {
	    fprintf(stderr, " (%s)", service_name) ;
	}
	fprintf(stderr, "\n") ;
    }

    /* server loop */
    do {
	struct sockaddr_in sa ;
	    
	alen = sizeof(sa) ;

	/* accept a connection */
	if ((active_socket = accept(socket_handle,
			  (struct sockaddr *) &sa,
			  &alen)) == -1) {
	    perror2("accept") ;
	} else {
	    /* if verbose, get name of peer and give message */
	    if (verboseflag) {
		struct hostent *he ;
		long norder ;
		char dotted[20] ;

		he = gethostbyaddr((char *)&sa.sin_addr.s_addr,
				   sizeof(sa.sin_addr.s_addr), AF_INET) ;
		if (!he) {
		    norder = htonl(sa.sin_addr.s_addr) ;
		    sprintf(dotted, "%d.%d.%d.%d",
			    (int)((norder >> 24) & 0xff),
			    (int)((norder >> 16) & 0xff),
			    (int)((norder >>  8) & 0xff),
			    (int)(norder & 0xff)) ;
		}
		fprintf(stderr, "connection from %s\n",
			(he ? he->h_name : dotted)) ;
	    }
	    if (forkflag) {
		switch (fork()) {
		  case 0:
		    handle_server_connection() ;
		    exit(0) ;
		  case -1:
		    perror2("fork") ;
		    break ;
		  default:
		    close(active_socket) ;
		    wait_for_children() ;
		}
	    } else {
		handle_server_connection() ;
	    }
	}
    } while (loopflag) ;
}


void handle_server_connection()
{
    /* open pipes to program, if requested */
    if (pipe_program != NULL) {
	open_pipes(pipe_program) ;
    }
    /* enter IO loop */
    do_io() ;
    /* connection is closed now */
    close(active_socket) ;
    if (pipe_program) {
	/* remove zombies */
	wait_for_children() ;
    }
}


void client(host, port, service_name)
char *host ;
int port ;
char *service_name ;
{
    /* get connection */
    active_socket = create_client_socket(&host, port) ;
    if (active_socket == -1) {
	perror2("client socket") ;
	exit(errno) ;
    } else if (active_socket == -2) {
	fprintf(stderr, "%s: unknown host %s\n", progname, host) ;
	exit(13) ;
    }
    if (verboseflag) {
	fprintf(stderr, "connected to %s port %d", host, port) ;
	if (service_name) {
	    fprintf(stderr, " (%s)", service_name) ;
	}
	fprintf(stderr, "\n") ;
    }

    /* open pipes to program if requested */
    if (pipe_program != NULL) {
	open_pipes(pipe_program) ;
    }
    /* enter IO loop */
    do_io() ;
    /* connection is closed now */
    close(active_socket) ;
}

/*EOF*/
